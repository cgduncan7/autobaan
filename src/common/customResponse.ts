import {
	Injectable,
	NestInterceptor,
	ExecutionContext,
	CallHandler,
} from '@nestjs/common'
import { Observable } from 'rxjs'
import { map } from 'rxjs/operators'

export interface CustomResponse<T = unknown> {
	data: T
}

@Injectable()
export class CustomResponseTransformInterceptor<T>
	implements NestInterceptor<T, CustomResponse<T>>
{
	intercept(
		_context: ExecutionContext,
		next: CallHandler,
	): Observable<CustomResponse<T>> {
		return next.handle().pipe(map((data) => ({ data })))
	}
}
