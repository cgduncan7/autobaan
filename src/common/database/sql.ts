export const TABLE_reservations = `
CREATE TABLE reservations (
  id                INT unsigned NOT NULL PRIMARY KEY AUTO_INCREMENT,
  username          VARCHAR(64) NOT NULL UNIQUE,
  password          VARCHAR(255) NOT NULL
  date_range_start  DATETIME NOT NULL,
  date_range_end    DATETIME NOT NULL,
  opponent_id       VARCHAR(32) NOT NULL,
  opponent_name     VARCHAR(255) NOT NULL
);
`
