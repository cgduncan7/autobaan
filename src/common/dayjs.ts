import dayjs from 'dayjs'
import isSameOrBefore from 'dayjs/plugin/isSameOrBefore'
import utc from 'dayjs/plugin/utc'
import timezone from 'dayjs/plugin/timezone'
import 'dayjs/locale/nl'
import { asyncLocalStorage } from './logger'

dayjs.extend(isSameOrBefore)
dayjs.extend(utc)
dayjs.extend(timezone)
dayjs.locale('nl')

dayjs.tz.setDefault('Europe/Amsterdam')

const dayjsTz = (
  date?: string | number | Date | dayjs.Dayjs | null | undefined
) => {
  asyncLocalStorage.getStore()?.debug(`<dayjs> timezoning ${date}`)
  return dayjs(date).tz()
}

export default dayjsTz
