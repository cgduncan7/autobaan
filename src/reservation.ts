export interface Time {
  hour: number
  minute: number
}

export interface Opponent {
  id: string
  name: string
}

export interface DateTime {
  year: number
  month: number
  day: number
  timeRange: {
    start: Time
    end: Time
  }
}

export const timeToString = ({ hour, minute }: Time): string =>
  `${`${hour}`.padStart(2, '0')}:${`${minute}`.padStart(2, '0')}`

export class Reservation {
  public readonly dateTime: DateTime
  public readonly opponent: Opponent
  public readonly possibleTimes: Time[]
  public booked = false

  constructor(dateTime: DateTime, opponent: Opponent) {
    this.dateTime = dateTime
    this.opponent = opponent
    this.possibleTimes = this.createPossibleTimes()
  }

  private createPossibleTimes() {
    const possibleTimes: Time[] = []

    const { start, end } = this.dateTime.timeRange

    let { hour, minute } = start
    const { hour: endHour, minute: endMinute } = end

    while (hour <= endHour && minute <= endMinute) {
      possibleTimes.push({ hour, minute })
      minute = (minute + 15) % 60
      if (minute === 0) {
        hour++
      }
    }

    return possibleTimes
  }
}
