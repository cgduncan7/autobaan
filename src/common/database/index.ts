import { resolve } from 'path'
import sqlite from 'sqlite3'
import { asyncLocalStorage } from '../logger'
import { CREATE_TABLE_reservations } from './sql'

export const run = async (sql: string, params?: unknown) => {
  const db = new sqlite.Database(resolve('autobaan_db'))
  await new Promise<void>((res, rej) => {
    asyncLocalStorage.getStore()?.debug(`<database> run ${sql} (${params})`)
    db.run(sql, params, (err) => {
      if (err) rej(err)
      res()
    })
  })
  db.close()
}

export const all = async <T>(sql: string, params?: unknown) => {
  const db = new sqlite.Database(resolve('autobaan_db'))
  const rows = await new Promise<T[]>((res, rej) => {
    asyncLocalStorage.getStore()?.debug(`<database> all ${sql} (${params})`)
    db.all(sql, params, (err, rows) => {
      if (err) rej(err)
      res(rows)
    })
  }
  )
  db.close()
  return rows
}

export const init = async () => {
  try {
    await run(CREATE_TABLE_reservations)
  } catch (err: unknown) {
    console.error(err)
  }
}

export class DatabaseError extends Error {}
export class DatabaseEnvironmentError extends DatabaseError {}
