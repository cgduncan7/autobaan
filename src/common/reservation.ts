import { Dayjs } from 'dayjs'
import { v4 } from 'uuid'
import dayjs from './dayjs'
import { all, run } from './database'

const RESERVATION_AVAILABLE_WITHIN_DAYS = 7

export interface User {
  username: string
  password: string
}

export interface Opponent {
  id: string
  name: string
}

export interface DateRange {
  start: Dayjs
  end: Dayjs
}

export class Reservation {
  public readonly id: string
  public readonly user: User
  public readonly dateRange: DateRange
  public readonly opponent: Opponent
  public readonly possibleDates: Dayjs[]
  public booked = false

  constructor(
    user: User,
    dateRange: DateRange,
    opponent: Opponent,
    possibleDates?: Dayjs[],
    id = v4()
  ) {
    this.id = id
    this.user = user
    this.dateRange = dateRange
    this.opponent = opponent
    this.possibleDates = possibleDates || this.createPossibleDates()
  }

  private createPossibleDates(): Dayjs[] {
    const possibleDates: Dayjs[] = []

    const { start, end } = this.dateRange

    let possibleDate = dayjs(start).second(0).millisecond(0)
    while (possibleDate.isSameOrBefore(end)) {
      possibleDates.push(possibleDate)
      possibleDate = possibleDate.add(15, 'minute')
    }

    return possibleDates
  }

  /**
   * Method to check if a reservation is available for reservation in the system
   * @returns is reservation date within 7 days
   */
  public isAvailableForReservation(): boolean {
    return (
      Math.ceil(this.dateRange.start.diff(dayjs(), 'days', true)) <=
      RESERVATION_AVAILABLE_WITHIN_DAYS
    )
  }

  public getAllowedReservationDate(): Dayjs {
    return this.dateRange.start
      .hour(0)
      .minute(0)
      .second(0)
      .millisecond(0)
      .subtract(RESERVATION_AVAILABLE_WITHIN_DAYS, 'days')
  }

  public toString() {
    return JSON.stringify(this.format())
  }

  public format() {
    return {
      user: {
        username: this.user.username,
        password: this.user.password,
      },
      opponent: this.opponent,
      booked: this.booked,
      possibleDates: this.possibleDates.map((date) => date.format()),
      dateRange: {
        start: this.dateRange.start.format(),
        end: this.dateRange.end.format(),
      },
    }
  }

  public serializeToJson(): SerializedReservation {
    return {
      user: this.user,
      opponent: this.opponent,
      booked: this.booked,
      possibleDates: this.possibleDates.map((date) => date.format()),
      dateRange: {
        start: this.dateRange.start.format(),
        end: this.dateRange.end.format(),
      },
    }
  }

  public static deserializeFromJson(
    serializedData: SerializedReservation
  ): Reservation {
    const start = dayjs(serializedData.dateRange.start)
    const end = dayjs(serializedData.dateRange.end)
    return new Reservation(
      serializedData.user,
      { start, end },
      serializedData.opponent,
      Reservation.deserializePossibleDates(serializedData.possibleDates)
    )
  }

  public static deserializePossibleDates(dates: string[]): Dayjs[] {
    return dates.map((date) => dayjs(date))
  }

  public static async save(res: Reservation) {
    await run(
      `
      INSERT INTO reservations
      (
        id,
        username,
        password,
        date_range_start,
        date_range_end,
        opponent_id,
        opponent_name
      )
      VALUES
      (
        ?,
        ?,
        ?,
        ?,
        ?,
        ?,
        ?
      )
    `,
      [
        res.id,
        res.user.username,
        res.user.password,
        res.dateRange.start.format('YYYY-MM-DD HH:mm:ss'),
        res.dateRange.end.format('YYYY-MM-DD HH:mm:ss'),
        res.opponent.id,
        res.opponent.name,
      ]
    )
  }

  public static async delete(res: Reservation) {
    await run(
      `
      DELETE FROM reservations
      WHERE id = ?
    `,
      [res.id]
    )
  }

  public static async fetchById(id: string): Promise<Reservation | null> {
    const response = await all<SqlReservation>(
      `
      SELECT *
      FROM reservations
      WHERE id = ?;
    `,
      [id]
    )

    if (response.length === 1) {
      const sqlReservation = response[0]
      const res = new Reservation(
        {
          username: sqlReservation.username,
          password: sqlReservation.password,
        },
        {
          start: dayjs(sqlReservation.date_range_start),
          end: dayjs(sqlReservation.date_range_end),
        },
        { id: sqlReservation.opponent_id, name: sqlReservation.opponent_name },
        undefined,
        sqlReservation.id
      )
      return res
    }

    return null
  }

  public static async fetchFirst(): Promise<Reservation | null> {
    const response = await all<SqlReservation>(
      `
      SELECT *
      FROM reservations
      ORDER BY date_range_start DESC
      LIMIT 1;
      `
    )

    if (response.length === 1) {
      const sqlReservation = response[0]
      const res = new Reservation(
        {
          username: sqlReservation.username,
          password: sqlReservation.password,
        },
        {
          start: dayjs(sqlReservation.date_range_start),
          end: dayjs(sqlReservation.date_range_end),
        },
        { id: sqlReservation.opponent_id, name: sqlReservation.opponent_name },
        undefined,
        sqlReservation.id
      )
      return res
    }

    return null
  }

  public static async fetchByDate(
    date: Dayjs,
    limit = 20
  ): Promise<Reservation[]> {
    const response = await all<SqlReservation>(
      `
      SELECT *
      FROM reservations
      WHERE DATE_FORMAT(DATE_SUB(date_range_start, INTERVAL 7 DAY), '%Y-%m-%d) = ?
      ORDER BY date_range_start DESC
      LIMIT ?;
      `,
      [date.format('YYYY-MM-DD'), limit]
    )

    if (response.length > 0) {
      return response.map(
        (sqlReservation) =>
          new Reservation(
            {
              username: sqlReservation.username,
              password: sqlReservation.password,
            },
            {
              start: dayjs(sqlReservation.date_range_start),
              end: dayjs(sqlReservation.date_range_end),
            },
            {
              id: sqlReservation.opponent_id,
              name: sqlReservation.opponent_name,
            },
            undefined,
            sqlReservation.id
          )
      )
    }

    return []
  }
}

export interface SerializedDateRange {
  start: string
  end: string
}

export interface SerializedReservation {
  user: User
  opponent: Opponent
  booked: boolean
  possibleDates: string[]
  dateRange: SerializedDateRange
}

export interface SqlReservation {
  id: string
  username: string
  password: string
  date_range_start: string
  date_range_end: string
  opponent_id: string
  opponent_name: string
}
