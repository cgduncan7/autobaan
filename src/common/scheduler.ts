import { Dayjs } from 'dayjs'
import { v4 } from 'uuid'

import { Logger, LogLevel } from './logger'
import { Reservation } from './reservation'
import { validateJSONRequest } from './request'
import { reserve } from './reserver'

export interface ScheduledReservation {
  reservation: Reservation
  scheduledFor?: Dayjs
}

export interface SchedulerResult {
  scheduledReservation?: ScheduledReservation
}

export type SchedulerInput = Record<string, unknown>

export const schedule = async (
  payload: SchedulerInput
): Promise<SchedulerResult> => {
  Logger.instantiate('scheduler', v4(), LogLevel.DEBUG)

  Logger.debug('Handling reservation', { payload })
  let reservation: Reservation
  try {
    reservation = await validateJSONRequest(payload)
  } catch (err) {
    Logger.error('Failed to validate request', { err })
    throw err
  }

  Logger.debug('Successfully validated request', {
    reservation: reservation.format(),
  })

  if (!reservation.isAvailableForReservation()) {
    Logger.debug(
      'Reservation date is more than 7 days away; saving for later reservation'
    )

    await Reservation.save(reservation)

    return {
      scheduledReservation: {
        reservation,
        scheduledFor: reservation.getAllowedReservationDate(),
      },
    }
  }

  Logger.info('Reservation request can be performed now')
  await reserve(reservation)
  return {
    scheduledReservation: { reservation },
  }
}
