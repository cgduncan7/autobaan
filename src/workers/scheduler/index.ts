import { Dayjs } from 'dayjs'
import { v4 } from 'uuid'

import { Logger, LogLevel } from '../../common/logger'
import { Reservation } from '../../common/reservation'
import { ReservationRequest, validateJSONRequest } from '../../common/request'
import { scheduleDateToRequestReservation } from '../../common/schedule'
import { Worker } from '../types'

export interface ScheduledReservationRequest {
  reservationRequest: ReservationRequest
  scheduledFor?: Dayjs
}

export interface ReservationSchedulerResult {
  scheduledReservationRequest?: ScheduledReservationRequest
}

export interface ReservationSchedulerInput
  extends Omit<ReservationRequest, 'dateRange'> {
  dateRange: { start: string; end: string }
}

export const work: Worker<ReservationSchedulerInput, ReservationSchedulerResult> = async (
  payload: ReservationSchedulerInput,
): Promise<ReservationSchedulerResult> => {
  Logger.instantiate('reservationScheduler', v4(), LogLevel.DEBUG)
  Logger.debug('Handling reservation', { payload })
  let reservationRequest: ReservationRequest
  try {
    reservationRequest = validateJSONRequest(payload)
  } catch (err) {
    Logger.error('Failed to validate request', { err })
    throw err
  }

  Logger.debug('Successfully validated request', { reservationRequest })

  const res = new Reservation(
    reservationRequest.dateRange,
    reservationRequest.opponent
  )

  if (!res.isAvailableForReservation()) {
    Logger.debug('Reservation date is more than 7 days away')
    const scheduledDay = scheduleDateToRequestReservation(
      reservationRequest.dateRange.start
    )
    Logger.info(
      `Scheduling reservation request for ${scheduledDay.format('YYYY-MM-DD')}`
    )
    return {
      scheduledReservationRequest: {
        reservationRequest,
        scheduledFor: scheduledDay,
      },
    }
  }

  Logger.info('Reservation request can be performed now')
  return {
    scheduledReservationRequest: { reservationRequest },
  }
}
