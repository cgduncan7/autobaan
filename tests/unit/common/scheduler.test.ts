import dayjs from '../../../src/common/dayjs'
import {
  ValidationError,
  ValidationErrorCode,
} from '../../../src/common/request'
import { Reservation } from '../../../src/common/reservation'
import { schedule, SchedulerInput } from '../../../src/common/scheduler'
import * as database from '../../../src/common/database'

jest.mock('../../../src/common/logger')
jest.mock('../../../src/common/reserver')
jest.mock('uuid', () => ({ v4: () => '1234' }))
jest.useFakeTimers().setSystemTime(new Date('2022-01-01'))

describe('scheduler', () => {
  test('should handle valid requests within reservation window', async () => {
    jest.spyOn(database, 'run').mockResolvedValueOnce()
    const start = dayjs().add(15, 'minutes')
    const end = start.add(15, 'minutes')

    const payload: SchedulerInput = {
      username: 'collin',
      password: 'password',
      dateRange: { start: start.toISOString(), end: end.toISOString() },
      opponent: { id: '123', name: 'collin' },
    }

    expect(await schedule(payload)).toMatchSnapshot({
      scheduledReservation: {
        reservation: {
          id: expect.any(String),
          user: {
            username: 'collin',
            password: expect.any(String),
          },
          dateRange: { start, end },
          opponent: { id: '123', name: 'collin' },
        },
      },
    })
  })

  test('should handle valid requests outside of reservation window', async () => {
    const start = dayjs().add(15, 'days')
    const end = start.add(15, 'minutes')
    const payload: SchedulerInput = {
      username: 'collin',
      password: 'password',
      dateRange: { start: start.toISOString(), end: end.toISOString() },
      opponent: { id: '123', name: 'collin' },
    }

    await expect(await schedule(payload)).toMatchSnapshot({
      scheduledReservation: {
        reservation: new Reservation(
          { username: 'collin', password: expect.any(String) },
          { start, end },
          { id: '123', name: 'collin' },
          undefined,
          '1234'
        ),
        scheduledFor: start
          .subtract(7, 'days')
          .hour(0)
          .minute(0)
          .second(0)
          .millisecond(0),
      },
    })
  })

  test('should throw error for invalid requests', async () => {
    const start = dayjs().add(15, 'days')
    const end = start.add(15, 'minutes')

    const payload: SchedulerInput = {
      password: 'password',
      dateRange: { start: start.toISOString(), end: end.toISOString() },
      opponent: { id: '123', name: 'collin' },
    }

    await expect(schedule(payload)).rejects.toThrowError(
      new ValidationError(
        'Invalid request',
        ValidationErrorCode.INVALID_REQUEST_BODY
      )
    )
  })
})
