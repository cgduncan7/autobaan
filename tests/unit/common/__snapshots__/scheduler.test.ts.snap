// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scheduler should handle valid requests outside of reservation window 1`] = `
{
  "scheduledReservation": {
    "reservation": {
      "booked": false,
      "dateRange": {
        "end": {
          "$D": 16,
          "$H": 1,
          "$L": "nl",
          "$M": 0,
          "$W": 0,
          "$d": {},
          "$m": 15,
          "$ms": 0,
          "$offset": 60,
          "$s": 0,
          "$u": false,
          "$x": {
            "$timezone": "Europe/Amsterdam",
          },
          "$y": 2022,
        },
        "start": {
          "$D": 16,
          "$H": 1,
          "$L": "nl",
          "$M": 0,
          "$W": 0,
          "$d": {},
          "$m": 0,
          "$ms": 0,
          "$offset": 60,
          "$s": 0,
          "$u": false,
          "$x": {
            "$timezone": "Europe/Amsterdam",
          },
          "$y": 2022,
        },
      },
      "id": "1234",
      "opponent": {
        "id": "123",
        "name": "collin",
      },
      "possibleDates": [
        {
          "$D": 16,
          "$H": 1,
          "$L": "nl",
          "$M": 0,
          "$W": 0,
          "$d": {},
          "$m": 0,
          "$ms": 0,
          "$offset": 60,
          "$s": 0,
          "$x": {
            "$timezone": "Europe/Amsterdam",
          },
          "$y": 2022,
        },
        {
          "$D": 16,
          "$H": 1,
          "$L": "nl",
          "$M": 0,
          "$W": 0,
          "$d": {},
          "$m": 15,
          "$ms": 0,
          "$offset": 60,
          "$s": 0,
          "$x": {
            "$timezone": "Europe/Amsterdam",
          },
          "$y": 2022,
        },
      ],
      "user": {
        "password": Any<String>,
        "username": "collin",
      },
    },
    "scheduledFor": {
      "$D": 9,
      "$H": 0,
      "$L": "nl",
      "$M": 0,
      "$W": 0,
      "$d": {},
      "$m": 0,
      "$ms": 0,
      "$offset": 60,
      "$s": 0,
      "$x": {
        "$timezone": "Europe/Amsterdam",
      },
      "$y": 2022,
    },
  },
}
`;

exports[`scheduler should handle valid requests within reservation window 1`] = `
{
  "scheduledReservation": {
    "reservation": {
      "booked": false,
      "dateRange": {
        "end": {
          "$D": 1,
          "$H": 1,
          "$L": "nl",
          "$M": 0,
          "$W": 6,
          "$d": {},
          "$m": 30,
          "$ms": 0,
          "$offset": 60,
          "$s": 0,
          "$u": false,
          "$x": {
            "$timezone": "Europe/Amsterdam",
          },
          "$y": 2022,
        },
        "start": {
          "$D": 1,
          "$H": 1,
          "$L": "nl",
          "$M": 0,
          "$W": 6,
          "$d": {},
          "$m": 15,
          "$ms": 0,
          "$offset": 60,
          "$s": 0,
          "$u": false,
          "$x": {
            "$timezone": "Europe/Amsterdam",
          },
          "$y": 2022,
        },
      },
      "id": Any<String>,
      "opponent": {
        "id": "123",
        "name": "collin",
      },
      "possibleDates": [
        "2022-01-01T00:15:00.000Z",
        "2022-01-01T00:30:00.000Z",
      ],
      "user": {
        "password": Any<String>,
        "username": "collin",
      },
    },
  },
}
`;
